Allow modifying the environment variables of a Builder command process

tareksha opened the pull request:
An extension might require setting/modifying the environment variables of the forked process that execute the command. Currently, the Che builder does not modify the environment of the built process at all and all the environment variables of the parent java process are passed as-is. Not only does this cause some issue due to possibly missing environment variables, but it also might pose a security issue due to exposing "too much" from the environment of the parent java process.

This change adds a an optional method that can be overridden to manipulate the environment variables that are set for the command process that is executed by Che builders. It also upgrades the codes slightly to use Java 8's new native Process timing APIs.

Signed-off-by: Tareq Sharafy tareq.sha@gmail.com

tareksha commented:
@sfroee
@skabashnyuk

skabashnyuk commented:
are you sure this is backward compatible change?

sfroee commented on platform-api/che-core-api-builder/src/main/java/org/eclipse/che/api/builder/internal/Builder.java:
```diff
@@ -486,6 +487,15 @@ public void cancelled(Cancellable cancellable) {
+	 * @param environmentVars The environment variables of the command process.
+	 * @param configuration Connfiguration of the build task.
+	 */
+	protected void modifyCommandEnvironment(Map<String, String> environmentVars, BuilderConfiguration configuration) {
```
good change! 
can we have a logger passed to this method as well?  so we can comment any change in the build log

tareksha replied:
yup, i'll add it @sfroee

sfroee commented on platform-api/che-core-api-builder/src/main/java/org/eclipse/che/api/builder/internal/Builder.java:
```diff
@@ -444,6 +444,7 @@ public Boolean call() throws Exception {
                     ProcessBuilder processBuilder = new ProcessBuilder().command(commandLine.toShellCommand()).directory(
                             configuration.getWorkDir()).redirectErrorStream(true);
+                    modifyCommandEnvironment(processBuilder.environment(), configuration);
                     Process process = processBuilder.start();
```
processBuilder.start() - need to be wrapped by some timer , so we can know the actual time spent in the command execution. - can be in another future change

tareksha replied:
I suggest using java 8's built-in waitFor(time, timeUnit) instead of using ad-hoc polling code. this areas is a candidate for several poll requests I sugess

tareksha commented:
@skabashnyuk yes it is, the added method has a default implementation and does not affect existing builder implementations.

tareksha committed:
Builder command exectuion extensibility & fixes

1- Use Java 8's standard process timing and killing methods
2- Allow customizing the ProcessBuilder before execution
3- Allow customizing source download & process wait
4- Print timeout terminate message AFTER full output
5- Make StreamPump implement AutoCloseable for easier use

Change-Id: I9d543cb8d9bdb19646cc62490b662356188431a2

tareksha commented:
@skabashnyuk  @sfroee 
the change has been updated to allow even more extensibility, plus using java 8's new standard Process API

skabashnyuk commented:
ok. @vparfonov @evoevodin ?

skabashnyuk commented:
@tareqhs  you can merge this pr
